### PR TITLE
Limit total partitions

### PR DESCRIPTION
**Pull request by knightXun**
because we leave limited memory and cpu for nebula in pubilc cloud, we must limit the number of spaces and partitions. Otherwise it's very dangerous. consider trial-version take one storaged, this may cause data loss. Never miss this

### CODE REVIEW

**Comment by knightXun** [comment]
If we dont limit the total number of partitions, the cpu and memory usage will very high.
image

**Comment by knightXun** [comment]
cc @dangleptr @yixinglu

**Comment by Shylock-Hg** [comment]
Please add test cases and check the error message.

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp [hunk 1]
```diff
@@ -119,6 +129,35 @@ CreateSpaceProcessor::pickHosts(PartitionID partId,
     return pickedHosts;
 }
 
+int32_t CreateSpaceProcessor::GetPartsNumbers() {
+    int32_t count = 0;
+    auto prefix = MetaServiceUtils::spacePrefix();
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        return -1;
```
Comment (laura-ding) [inline src/meta/processors/partsMan/CreateSpaceProcessor.cpp:138]: I don't know what it means to return -1 here. the caller does not process it. 
If here failed, is means the rocksdb may be abnormal, and we should return and don’t continue.

File: src/interface/meta.thrift [hunk 2]
```diff
@@ -49,6 +50,7 @@ enum ErrorCode {
     E_INVALID_CHARSET           = -45,
     E_INVALID_COLLATE           = -46,
     E_CHARSET_COLLATE_NOT_MATCH = -47,
+    E_TOO_LARGE_REPLICA_FACTOR   = -48,
```
[the remark this answers is not in the document]
Comment (knightXun) [reply src/interface/meta.thrift:53]: good point

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp (hunk 1 above)
Comment (knightXun) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp:138]: If return -1, we should throw a exception here?

Comment (knightXun) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp:144]: fix this

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp [hunk 3]
```diff
@@ -119,6 +129,35 @@ CreateSpaceProcessor::pickHosts(PartitionID partId,
...
+    }
+
+    while (iter->valid()) {
+        auto spaceId_ = MetaServiceUtils::spaceId(iter->key());
```
[the remark this answers is not in the document]
Comment (knightXun) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp]: fixed

**Comment by knightXun** [comment]
> Please add test cases and check the error message.

done

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp [hunk 4]
```diff
@@ -119,6 +134,38 @@ CreateSpaceProcessor::pickHosts(PartitionID partId,
...
+        auto spaceName = MetaServiceUtils::spaceName(iter->val());
+        std::unique_ptr<kvstore::KVIterator> partIter;
+        auto kvRet = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, partPrefix, &partIter);
+        if ( kvRet != kvstore::ResultCode::SUCCEEDED ) {
+            LOG(ERROR) << "List Partitions From " << spaceName
+                       << ", Space id " << spaceId << " Failed";
+            return -1;
+        }
+
+        while (partIter->valid()) {
```
Comment (laura-ding) [inline src/meta/processors/partsMan/CreateSpaceProcessor.cpp:159]: You can get the partition_num from space properties, don't need list parts.

Comment (knightXun) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp]: good point, Done

**Comment by knightXun** [comment]
cc @laura-ding done

File: src/graph/test/PartitionTest.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
Comment (laura-ding) [inline src/graph/test/PartitionTest.cpp:1]: 2020

File: src/graph/test/PartitionTest.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "base/Base.h"
+#include "graph/test/TestEnv.h"
+#include "graph/test/TestBase.h"
+#include "meta/test/TestUtils.h"
+#include "storage/test/TestUtils.h"
+
+DECLARE_int32(heartbeat_interval_secs);
```
Comment (laura-ding) [inline src/graph/test/PartitionTest.cpp:13]: Don't need

File: src/graph/test/PartitionTest.cpp (hunk 5 above)
Comment (knightXun) [reply src/graph/test/PartitionTest.cpp]: done

File: src/graph/test/PartitionTest.cpp (hunk 6 above)
Comment (knightXun) [reply src/graph/test/PartitionTest.cpp]: done

**Comment by knightXun** [comment]
cc @dutor

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp [hunk 7]
```diff
@@ -9,6 +9,7 @@
 
 DEFINE_int32(default_parts_num, 100, "The default number of parts when a space is created");
 DEFINE_int32(default_replica_factor, 1, "The default replica factor when a space is created");
+DEFINE_int32(default_max_parts_num, 100, "The default max partitions number");
```
Comment (Shylock-Hg) [inline src/meta/processors/partsMan/CreateSpaceProcessor.cpp:12]: I think you can set default to max, and give it a property value in configuration when deploy.

Comment (knightXun) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp:12]: 100 partitions is enough for trial version, user can create up to 100 spaces.

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp [hunk 8]
```diff
@@ -119,6 +134,24 @@ CreateSpaceProcessor::pickHosts(PartitionID partId,
     return pickedHosts;
 }
 
+int32_t CreateSpaceProcessor::GetPartsNumbers() {
```
Comment (wadeliuyi) [inline src/meta/processors/partsMan/CreateSpaceProcessor.cpp:137]: Is this function mean get all part number of all spaces?

Comment (knightXun) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp:137]: yeap

File: src/meta/processors/partsMan/CreateSpaceProcessor.cpp (hunk 7 above)
Comment (dangleptr) [reply src/meta/processors/partsMan/CreateSpaceProcessor.cpp:12]: Currently,  the largest number for parts is 2^24,   the default max number should be it.

**Comment by dangleptr** [comment]
That's fine for cloud branch

**Pull request merged**